compiler/rustc_expand/src/base.rs

```diff
@@ -761,6 +763,75 @@ impl SyntaxExtension {
         }
     }
 
+    fn collapse_debuginfo_by_name(sess: &Session, attr: &Attribute) -> CollapseMacroDebuginfo {
+        use crate::errors::CollapseMacroDebuginfoIllegal;
+        // #[collapse_debuginfo] without enum value (#[collapse_debuginfo(no/external/yes)])
+        // considered as `yes`
+        attr.meta_item_list().map_or(CollapseMacroDebuginfo::Yes, |l| {
+            let [NestedMetaItem::MetaItem(_item)] = &l[..] else {
+                sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: attr.span });
+                return CollapseMacroDebuginfo::Unspecified;
+            };
+            let item = l[0].meta_item();
+            item.map_or(CollapseMacroDebuginfo::Unspecified, |mi| {
+                if !mi.is_word() {
+                    sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: mi.span });
+                    CollapseMacroDebuginfo::Unspecified
+                } else {
+                    match mi.name_or_empty() {
+                        sym::no => CollapseMacroDebuginfo::No,
+                        sym::external => CollapseMacroDebuginfo::External,
+                        sym::yes => CollapseMacroDebuginfo::Yes,
+                        _ => {
+                            sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: mi.span });
+                            CollapseMacroDebuginfo::Unspecified
+                        }
+                    }
+                }
+            })
+        })
+    }
+
+    /// if-ext - if macro from different crate (related to callsite code)
+    /// | cmd \ attr    | no  | (unspecified) | external | yes |
+    /// | no            | no  | no            | no       | no  |
+    /// | (unspecified) | no  | no            | if-ext   | yes |
+    /// | external      | no  | if-ext        | if-ext   | yes |
+    /// | yes           | yes | yes           | yes      | yes |
+    pub fn should_collapse(
+        flag: CollapseMacroDebuginfo,
+        attr: CollapseMacroDebuginfo,
+        ext: bool,
+    ) -> bool {
+        #[rustfmt::skip]
+        let collapse_table = [
+            [false, false, false, false],
+            [false, false, ext,   true],
+            [false, ext,   ext,   true],
+            [true,  true,  true,  true],
+        ];
+        let rv = collapse_table[flag as usize][attr as usize];
+        tracing::debug!(?flag, ?attr, ?ext, ?rv);
+        rv
+    }
+
+    fn get_collapse_debuginfo(
+        sess: &Session,
+        span: Span,
```

Comment:
This span is only used for logging.

(I'm also pretty skeptical about keeping debug logging in already debugged code, it will very rarely be useful in the future and it can be temporarily added again in those cases.)